File: app/modules/ytdlp/android/src/main/java/expo/modules/ytdlp/platforms/YouTube.kt
package expo.modules.ytdlp.platforms

/** yt-dlp's phrasing for YouTube's "Sign in to confirm you're not a bot" gate. */
private val SIGN_IN_GATE_PATTERN = Regex("sign in to confirm you.{1,2}re not a bot", RegexOption.IGNORE_CASE)

private fun isSignInGateError(message: String): Boolean = SIGN_IN_GATE_PATTERN.containsMatchIn(message)

class YouTube(url: String) : PlaylistCapableBasePlatform(url) {
    companion object {
        val HOSTS = setOf("youtube.com", "www.youtube.com", "m.youtube.com", "music.youtube.com", "youtu.be")
    }

    override fun defaultThumbnail(id: String): String = "https://i.ytimg.com/vi/$id/hqdefault.jpg"

    /** The known-permanent sign-in gate is never worth retrying; every other failure is treated as transient. */
    override fun isRetryableError(error: Throwable): Boolean = !isSignInGateError(describeErrorRaw(error))

    /**
     * yt-dlp surfaces YouTube's own "Sign in to confirm you're not a bot" gate verbatim, including
     * a raw stack of wiki links — not actionable for a user, since it requires real logged-in
     * cookies to bypass, not anything this app can retry or work around on its own.
     */
    override fun describeError(error: Throwable): Pair<String, Map<String, Any?>?> {
        val raw = describeErrorRaw(error)
        return if (isSignInGateError(raw)) "errors.signInRequired" to null else "errors.raw" to mapOf("raw" to raw)
    }
}

File: server/src/platforms/YouTube.ts
import { BasePlatform } from "./BasePlatform.js";
import { PlaylistCapableBasePlatform } from "./PlaylistCapableBasePlatform.js";
import type { UserFacingError } from "./Platform.js";

// yt-dlp surfaces YouTube's own "Sign in to confirm you're not a bot" gate verbatim, including a
// raw stack of wiki links — not actionable for a user, since it requires real logged-in cookies to
// bypass, not anything this app can retry or work around on its own. Only used for responses sent
// to the app; the technical message still goes into the job log file for debugging.
const SIGN_IN_GATE_PATTERN = /sign in to confirm you.{1,2}re not a bot/i;

export class YouTube extends PlaylistCapableBasePlatform {
  readonly id = "youtube";

  static readonly HOSTS = ["youtube.com", "www.youtube.com", "m.youtube.com", "music.youtube.com", "youtu.be"];

  private static readonly isSignInGateError = BasePlatform.permanentFailureMatcher(SIGN_IN_GATE_PATTERN);

  /** The known-permanent sign-in gate is never worth retrying; every other failure is treated as transient. */
  override isRetryableError(err: unknown): boolean {
    return !YouTube.isSignInGateError(err);
  }

  override describeError(err: unknown, fallbackRaw?: string): UserFacingError {
    return YouTube.isSignInGateError(err, fallbackRaw)
      ? { key: "errors.signInRequired" }
      : super.describeError(err, fallbackRaw);
  }

  override defaultThumbnail(entryId: string): string | null {
    return `https://i.ytimg.com/vi/${entryId}/hqdefault.jpg`;
  }
}
